File: test/csharp/SpecGateFixtures/Conformance/SumTypes/checked_divide.cs
using SpecGate.Annotations;

namespace SpecGateFixtures.Conformance.SumTypes;

/// <summary>
/// Fallible operation exercising precise <c>[SpecException]</c> resolution: the
/// declared <see cref="DivideByZeroException"/> is the <c>Err</c> arm, while an
/// undeclared <see cref="InvalidOperationException"/> falls through to
/// <c>$fault</c> (the panic analog).
/// </summary>
public static class CheckedDivideOps
{
    /// <summary>Divides <paramref name="a"/> by <paramref name="b"/>.</summary>
    /// <param name="a">The dividend (spec input <c>a</c>).</param>
    /// <param name="b">The divisor (spec input <c>b</c>).</param>
    /// <returns>
    /// The quotient (Ok arm). Throws the declared <see cref="DivideByZeroException"/>
    /// for the Err arm; an undeclared throw becomes <c>$fault</c>.
    /// </returns>
    [SpecOperation("checked_divide", Spec = "fixture.checked_divide")]
    [SpecException(typeof(DivideByZeroException))]
    public static int CheckedDivide([SpecInput("a")] int a, [SpecInput("b")] int b)
    {
        if (b == 0)
        {
            throw new DivideByZeroException("division by zero");
        }

        if (b < 0)
        {
            throw new InvalidOperationException("negative divisor");
        }

        return a / b;
    }
}

File: test/rust/crates/specgate-fixtures/src/conformance/sum_types/checked_divide.rs
// Fallible op with a declared error arm AND an undeclared panic path. Exercises
// precise `[SpecException]` resolution on the C# side: the declared exception
// maps to the Err arm, while an undeclared throw falls through to `$fault`. The
// Rust target realizes the same contract with `Result` plus `panic!`.
use specgate::*;

#[spec_operation("checked_divide", spec = "fixture.checked_divide")]
pub fn checked_divide(a: i32, b: i32) -> Result<i32, String> {
    if b == 0 {
        return Err("division by zero".to_string());
    }
    if b < 0 {
        panic!("negative divisor");
    }
    Ok(a / b)
}
